### PR TITLE
Add latency statistics accumulation to GlueHiveMetastore

### PR DESCRIPTION
**Pull request by alexjo2144**
These are used by the JMX connector to present stats similar to the ThriftMetastoreStats: https://github.com/prestosql/presto/blob/master/presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftMetastoreStats.java

Issue: https://github.com/prestosql/presto/issues/1294

~~There were a few calls I didn't add stats to because they do async api calls so the stats didn't seem to make as much sense. Those were~~
- ~~getPartitionsByNames~~
- ~~addPartitions~~

### CODE REVIEW

**Comment by aalbu** [comment]
Could you note which calls you didn't record stats for?

**Comment by aalbu** [comment]
I noticed that there aren't any tests for the other metastores, but could you try to add one that proves stats are being collected here?  Basically make a call and then select from the corresponding `jmx.current...` table to show that the call count has been incremented?

**Comment by alexjo2144** [comment]
Added one unit test which just looks at the stat accumulation, I'll try to add another integration test to test the hook in with JMX.

**Comment by alexjo2144** [comment]
I was able to make a full integration test between Glue and the JMX connector, but had to make a few changes. 

- Added jmx as a test dependency to presto-hive
- Added one entry to the main pom's `dependencyManagement` list. I'm not fully sure I understand what that does, but was needed for the first point.
- Allowed for a null metastore in the TestingHiveConnectorFactory
  - Needed this in order for the `hive.metastore` property to take effect

File: presto-hive/pom.xml [hunk 1]
```diff
@@ -328,6 +334,7 @@
                         <configuration>
                             <excludes>
                                 <exclude>**/TestHiveGlueMetastore.java</exclude>
+                                <exclude>**/TestGlueJMXStatistics.java</exclude>
```
Comment (alexjo2144) [inline presto-hive/pom.xml]: Looks like there's a `test-hive-glue` profile that got dropped when we moved from travis to github workflows?

**Comment by electrum** [comment]
Let’s not use the JMX connector for testing, as that adds complexity and is not necessary. Instead, we can simply ask the MBean server and access the stats directly.

File: presto-hive/pom.xml (hunk 1 above)
Comment (electrum) [reply presto-hive/pom.xml]: These were never actually used

**Comment by electrum** [comment]
I think this is missing the Guice binding to export stats. For now, we can drop the integration tests. Verify manually that the stats are being exported correctly when you run the full server.

**Comment by alexjo2144** [comment]
Sounds good, thanks David. I rolled back the last few commits with the integration test attempts. Here's a screenshot from the cli querying the stats via the jmx connector:

image

**Comment by findepi** [comment]
> There were a few calls I didn't add stats to because they do async api calls so the stats didn't seem to make as much sense. Those were
> 
> * getPartitionsByNames
> * addPartitions

I think we want to track these too. 

eg for one `getPartitionsByNames` call, we do multiple `batchGetPartitionAsync` requests.
We could even track both, but I think we can start from tracking `getPartitionsByNames`  invocations first.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/glue/GlueHiveMetastore.java [hunk 2]
```diff
@@ -245,17 +256,19 @@ private static AWSCredentialsProvider getCustomAWSCredentialsProvider(String pro
     public List<String> getAllDatabases()
     {
         try {
-            List<String> databaseNames = new ArrayList<>();
-            String nextToken = null;
+            return stats.getGetAllDatabases().wrap(() -> {
+                List<String> databases = new ArrayList<>();
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/glue/GlueHiveMetastore.java]: unrelated change (used to be `databaseNames`). This can stay, but generally please avoid unrelated changes

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/glue/GlueHiveMetastore.java [hunk 3]
```diff
@@ -822,27 +849,33 @@ public void dropColumn(HiveIdentity identity, String databaseName, String tableN
...
+                            .withTableName(tableName)
+                            .withPartitionInputList(partitionInputs)));
+                }
+
+                for (Future<BatchCreatePartitionResult> future : futures) {
+                    try {
+                        BatchCreatePartitionResult result = future.get();
+                        propagatePartitionErrorToPrestoException(databaseName, tableName, result.getErrors());
+                    }
+                    catch (InterruptedException e) {
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/glue/GlueHiveMetastore.java]: Add `throw new PrestoException(HIVE_METASTORE_ERROR, e);` to catch clause (after re-interrupting current thread)

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/glue/GlueMetastoreApiStats.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+@ThreadSafe
+public class GlueMetastoreApiStats
+{
+    private final TimeStat time = new TimeStat(MILLISECONDS);
+    private final CounterStat totalFailures = new CounterStat();
+
+    public <V, E extends Exception> V wrap(ThrowingCallable<V, E> callable)
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/glue/GlueMetastoreApiStats.java]: "wrap" should return a wrapping callable (`Callable<V>`)

rename to `call`

(you can fix `JdbcApiStats` in separate PR)

File: presto-hive/src/test/java/io/prestosql/plugin/hive/metastore/glue/TestHiveGlueMetastore.java [hunk 5]
```diff
@@ -113,4 +115,27 @@ public void testGetPartitions() throws Exception
...
+    public void testGetDatabasesLogsStats()
+    {
+        GlueHiveMetastore metastore = (GlueHiveMetastore) getMetastoreClient();
+        GlueMetastoreStats stats = metastore.getStats();
+        double initialCallCount = stats.getGetAllDatabases().getTime().getAllTime().getCount();
+        long initialFailureCount = stats.getGetAllDatabases().getTotalFailures().getTotalCount();
+        getMetastoreClient().getAllDatabases();
+        assertEquals(stats.getGetAllDatabases().getTime().getAllTime().getCount(), initialCallCount + 1.0);
+        assertTrue(stats.getGetAllDatabases().getTime().getAllTime().getAvg() > 0.0);
+        assertEquals(stats.getGetAllDatabases().getTotalFailures().getTotalCount(), initialFailureCount);
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/metastore/glue/TestHiveGlueMetastore.java]: Add `@Test(singleThreaded = true)` to the test class

**Comment by alexjo2144** [comment]
Fixed and rebased. Thanks Piotr

**Comment by findepi** [comment]
CI failed -- https://github.com/prestosql/presto/issues/2616

**Comment by findepi** [comment]
Merged as 7c32534449ba2416876e059373e5852d817d893c, thanks!

(i amended the commit message)
